Convert one h5 file per collection (4) to parquet

chuckwondo opened the pull request:
There's a fair bit of work in this PR, so the following summary should help orient/focus review.

Primary functions of interest:

- `to_polars` in `src/gedi_geoparquet/hdf5.py`: converts a GEDI hdf5 file to a Polars LazyFrame
- `to_arrow` in `src/gedi_geoparquet/hdf5.py`: converts a GEDI hdf5 file to an Arrow Table
- `scan_hdf5` in `src/gedi_geoparquet/polars_.py`: converts a GEDI hdf5 group to a Polars LazyFrame (called by `to_polars` to convert each GEDI beam group to a LazyFrame and then concatenate them vertically -- leverages Polars I/O plugin mechanism to minimize I/O code)

Additionally:

- All remaining functions are "internal", in support of the primary functions above.
- Files under `src/gedi_geoparquet/schema/` simply codify the schemas for the 4 GEDI collections. Each schema python file defines the _entirety_ of each collection's schema, so they are easily human readable, but only the datasets that we previously decided to include (see #2) are uncommented. All datasets we want to exclude are commented out, but left in place for reference (or in case we decide to change which to include/exclude).

Scripts:

- Additions to `README.md` should help understanding the new scripts.
- `scripts/convert.py`: CLI for converting a single GEDI hdf5 file to a single parquet file (based upon the relevant collection schema). Basically a CLI wrapper for the `to_polars` and `to_arrow` functions for reading and converting, and then writing to parquet.
- `scripts/join.py`: CLI for joining individual corresponding GEDI parquet files (generated from `convert.py`) into a singular parquet file (so that 4 individual, related files get joined on shot_number into a singular GEDI representation, where duplicate columns are dropped).

chuckwondo committed:
Support converting hdf5 to geoparquet.

Fixes #8

chuckwondo committed:
Set zstd as default compression for join script

kylebarron commented on src/gedi_geoparquet/schema/gedi_l4c.py:
```diff
@@ -0,0 +1,166 @@
+# The full schema was produced via gedi_geoparquet.polars_._infer_schema.
+# Although we're selecting only a subset of columns, we're leaving the unused
+# columns in comments so we have a full reference of what's available.
+SCHEMA: t.Final = pl.Schema(
```
I didn't realize that Polars's schema doesn't support exporting to Arrow 😬 

```py
import polars as pl
from arro3.core import Schema, DataType, Field

# This doesn't work to define the schema in polars and pass to arrow
schema = pl.Schema({"str": pl.String()})
Schema(schema)

# This works to define the schema in arro3 and pass to polars
arro3_schema = Schema([Field("str", DataType.string())])
Schema(arro3_schema)
```

So I think it would be better to define the schema using pyarrow or arro3-core and pass it to Polars as needed. Then the schema would be more general, not just tied to Polars.

This also means I need to make an issue/implement schema export to Arrow

kylebarron replied:
Sorry I was actually using my own API wrong 🤦‍♂️ 

This works:

```py
import polars as pl
from arro3.core import Schema

# This doesn't work to define the schema in polars and pass to arrow
schema = pl.Schema({"str": pl.String()})
Schema.from_arrow(schema)
```

chuckwondo replied:
Yes, I was back and forth on this.  I chose Polars because I could easily copy the repr of the Polars schema and with minimal typing/mods, produce this dict, whereas the PyArrow repr of a schema is far from looking like a dict, thus requiring a fair bit of typing on my part to produce the PyArrow equivalent of this Polars schema.

Also, there's no need to use `arro3.core.Schema` directly.  PyArrow.Schema and Polars.Schema both implement the pycapsule stuff, so both can be directly converted between each other.  The issue is that the Polars.Schema does not retain any field or schema metadata.  It retains only the field names and data types, so the conversion from an arrow schema to a Polars.Schema is lossy.  Metadata is lost on roundtripping.

However, having said all of that.  I'll go through the effort of making this a PyArrow Schema instead of a Polars Schema.  I agree that it will be better that way (I was just being a bit lazy).

kylebarron commented on src/gedi_geoparquet/hdf5.py:
```diff
@@ -0,0 +1,344 @@
+
+    # WARNING: When Polars converts to Arrow, it defaults to using LargeString and
+    # LargeList even when String and List will suffice.  This might be problematic.
+    # See https://github.com/pola-rs/polars/issues/15047.
```
This isn't true anymore I'm pretty sure. Polars uses the new "string view" type for all its data, which other Arrow implementations also support.

kylebarron commented on pyproject.toml:
```diff
@@ -29,7 +26,9 @@ license = "Apache-2.0"
 license-files = ["LICEN[CS]E*"]
 dependencies = [
     "geoarrow-pyarrow>=0.2.0",
+    "geoarrow-types>=0.3.0",
```
did you use this somewhere?

chuckwondo commented on scripts/convert.py:
```diff
@@ -0,0 +1,117 @@
+import pyarrow.parquet as pq
+from cyclopts import App
+
+type Compression = t.Literal["brotli", "gzip", "lz4", "snappy", "zstd"]
```
Yes. The `type` statement was introduced in 3.12.  See <https://peps.python.org/pep-0695/>.

chuckwondo commented on src/gedi_geoparquet/polars_.py:
I named it this way to avoid naming conflict with the `polars` library itself, but since I'm importing with aliasing anyway (`import gedi_geoparquet.polars_ as pl_`), I don't need to, so I will simply rename as you suggest.

chuckwondo committed:
Rename polars_.py to polars.py

chuckwondo committed:
Move geoarrow-types to dev dependencies

chuckwondo committed:
Move/rename polars._flatten to hdf5.flatten

kylebarron commented on scripts/join.py:
```diff
@@ -0,0 +1,75 @@
+    ## This doesn't seem to write the metadata where Arrow recognizes it, or at
+    ## least not where it will be attached to the Arrow Schema upon reading of
+    ## the schema.
+    # lf.sink_parquet(output, metadata=GEOPARQUET_METADATA)
```
> Rather, it is written _next to_ the arrow metadata, as a sibling metadata entry.

That is the correct location. GeoParquet defines its own `geo` entry

```python
import polars as pl
import pyarrow.parquet as pq

df = pl.DataFrame({"a": [1, 2, 3], "b": [4, 5, 6]})
df.write_parquet("tmp.parquet", metadata={"hello": "world"})
pq.read_metadata("tmp.parquet").metadata
# {b'hello': b'world',
#  b'ARROW:schema': b'/////6YAAAAEAAAA8v///xQAAAAEAAEAAAAKAAsACAAKAAQA+P///wwAAAAIAAgAAAAEAAIAAAA0AAAABAAAAMD///8gAAAAEAAAAAgAAAABAgAAAAAAALj///9AAAAAAQAAAAEAAABiAAAA7P///zgAAAAgAAAAGAAAAAECAAAQABIABAAQABEACAAAAAwAAAAAAPT///9AAAAAAQAAAAgACQAEAAgAAQAAAGEA'}
```

This is correct. 

Having the geo metadata inside of the Arrow schema is _ok_, but is separate from the GeoParquet spec.

chuckwondo committed:
Generate schema via pyarrow, not polars

chuckwondo commented:
@kylebarron, apologies for the volume of changes.

The **tl;dr** is that you can simply run `make join` to run the whole shebang to produce parquet files for you to examine, if you wish. (NOTE: this requires you to populate your `~/.netrc` file with your Earthdata Login creds, as it will read granule files via https).

- I did a fair bit of refactoring to leverage pyarrow for the schema rather than polars.
- I updated the scripts to account for the refactoring.
- I replaced argparse with cyclopts to greatly simplify the scripts. I really like cyclopts.

Running `make join` will do the following:

  - generate the full schema for each of the 4 GEDI collections by reading an arbitrarily chosen granule from collection (this _should_ be a noop, as they are already included as resources in the repo [the *.arrows files], unless pulling the latest updates from git causes timestamps to look like the *.arrows files are out of date)
  - read the 4 individual granules files (`.h5`) via https (hence the need for EDL creds in your .netrc), converting them to `.parquet` files (including only a subset of columns as specified in the `src/gedi_geoparquet/schema/gedi_*.py` files), writing them to the `data` directory (ignored by git)
  - join the 4 files on `shot_number`, writing the result to the `data` directory (I've also added some filtering to filter out rows that users that subset GEDI data currently always do themselves, so this reduces how much data we need to copy from the h5 archives)

chuckwondo committed:
Bump pre-commit-hook rev

chuckwondo committed:
Ensure output dir exists

chuckwondo committed:
Make schema param positional, no kw-only.

chuckwondo committed:
Drop "constant" columns

chuckwondo committed:
Add batched function to batch h5py Dataset

chuckwondo committed:
Remove obsolete/unused functions

chuckwondo committed:
Remove unnecessary use of ArrowSchemaExportable

chuckwondo committed:
Simplify logic in _basename

chuckwondo committed:
Move internal comments to better locations

chuckwondo committed:
Fix scan_hdf5 to properly handle batching
